Fix/translation-languages

Bluesmile82 opened the pull request:
This PR  moves the current locale through url param implementation to a next integration via 118n in next.config and prefixes to show on the URL

Also the locales are adapted to the ones existing in the backend
    
## Testing instructions

Check that the locales and language change is working correctly

## Tracking

https://vizzuality.atlassian.net/browse/RA2-159?atlOrigin=eyJpIjoiNmQ5MzQxN2Y2Mjk0NDBiNTkwZjgzOGU2YmZlMGQ5ZTciLCJwIjoiaiJ9

https://vizzuality.atlassian.net/browse/RA2-97?atlOrigin=eyJpIjoiZTQyM2Q2MThlMTY1NGM4YWE0NzMxNDBiMTdkMzQxNGIiLCJwIjoiaiJ9

Bluesmile82 commented:
I had some problems with the cypress tests with the i18n. It seems to get some uncaught:exceptions that are difficult to trace. Please let me know if you had experienced something similar. For now I'm just bypassing all errors for uncaught:exceptions but this could be improved

davidsingal reviewed (changes requested):
I think we should change how to show the locale in the URL, here more details: https://vizzuality.atlassian.net/browse/RA2-159?focusedCommentId=23099

Also, this PR could be pointed to https://github.com/ConservationInternational/resilienceatlas/pull/140

clementprdhomme commented on frontend/next.config.js:
```diff
@@ -30,6 +30,12 @@ const nextConfig = {
     ];
   },
+  i18n: {
+    locales: ['default', 'en', 'fr', 'es', 'zh-CN', 'pt-BR', 'ru'],
+    defaultLocale: 'default',
```
Can you explain what is this `default` locale? Why are we not defaulting to `en`?

Bluesmile82 replied:
This is just a trick on the Next documentation to add the 'en' prefix if we don't have any specified https://nextjs.org/docs/advanced-features/i18n-routing#prefixing-the-default-locale

clementprdhomme replied:
Ok, got it. Do we need to prefix the default locale though? If we don't, you don't have to update all the Cypress tests.

I think it's ok to go to the English version by visiting resilienceatlas.org and not resilienceatlas.org/en.

clementprdhomme commented on frontend/src/middleware.ts:
```diff
@@ -3,13 +3,33 @@ import { PORT } from 'state/utils/api';
+    return;
+  }
+
+  // Set the default locale (en) as a prefix if we don't have a locale in the url
+  if (request.nextUrl.locale === 'default') {
+    const locale = request.cookies.get('NEXT_LOCALE') || 'en';
+
+    return NextResponse.redirect(
+      new URL(`/${locale}${request.nextUrl.pathname}${request.nextUrl.search}`, request.url),
+    );
+  }
```
If we don't have the `default` locale, I don't think we need any of this. Next.js automatically redirect the user based on the detected locale: https://nextjs.org/docs/advanced-features/i18n-routing#automatic-locale-detection.

You might have some challenges with sending the locale to the API depending on the origin of the request (SSR or client). Here you can see how it was dealt with in HeCo Invest:

- https://github.com/Vizzuality/heco-invest/blob/develop/frontend/pages/_app.tsx#L60-L84
- https://github.com/Vizzuality/heco-invest/blob/develop/frontend/services/api.ts#L26-L44

Anyway, I don't think you need to work on this issue in this PR.

clementprdhomme commented on frontend/src/pages/_app.tsx:
```diff
@@ -94,12 +93,11 @@ const ResilienceApp = ({ Component, ...rest }: AppPropsWithLayout) => {
-    tx.setCurrentLocale(lang);
-  }, [lang]);
+    tx.setCurrentLocale(router.locale);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```
Any reason why you're not passing `router` in the list of dependencies?

clementprdhomme commented on frontend/src/views/components/LanguageSwitcher/LanguageSwitcher.component.tsx:
```diff
@@ -3,19 +3,20 @@ import React, { useMemo } from 'react';
 import cx from 'classnames';
 
-const AVAILABLE_LANGUAGES = ['en', 'fr', 'es', 'zh', 'pt', 'ru'];
+const AVAILABLE_LANGUAGES = ['en', 'fr', 'es', 'zh-CN', 'pt-BR', 'ru'];
```
Look at this HeCo Invest example: https://github.com/Vizzuality/heco-invest/blob/develop/frontend/locales.config.json.

We stored the locales in a separate top-level JSON and we imported it in the components and `next.config.js`. It avoids defining the locales in multiple places. It's more future-proof.

clementprdhomme commented on frontend/src/views/components/LanguageSwitcher/LanguageSwitcher.component.tsx:
```diff
@@ -27,14 +28,22 @@ type Language = {
-    setParam('lang', code);
+  const router = useRouter();
+
+  const setLanguage = (locale) => {
+    const { pathname, asPath, query } = router;
+    // change just the locale and maintain all other route information including href's query
+    router.push({ pathname, query }, asPath, { locale });
+  };
```
Can you set the cookie as well so that the user's language preference is kept in the future?

See an example here: https://github.com/Vizzuality/heco-invest/blob/develop/frontend/containers/layouts/language-selector/component.tsx#L27-L39

clementprdhomme reviewed (commented):
I agree with David this PR should be merged to `feature/static-translations` instead of `develop`.

I disagree with David about using query parameters to store the locale in the URL. If you're concerned about the SEO, this is what you should do to tell search engines how pages are related to each other: https://github.com/Vizzuality/heco-invest/blob/develop/frontend/components/head/component.tsx#L34-L46. @Bluesmile82 I strongly suggest that you add these tags on all the pages.

Bluesmile82 commented:
@clementprdhomme  Ready again: 

- Added the cookie for the user preference
- Added the locales config json file
- Removed the 'default' locale
- Added router to missing dependencies

clementprdhomme reviewed (commented):
A few more comments. We're close to finishing with this task 😉.

Bluesmile82 commented:
Thanks for the comments. Ready again. I simplify the code relying on next and the locale config as the source so at the end we don't depend on transifex for the locales.

clementprdhomme reviewed (approved):
Thanks Álvaro for your work! It looks good!
